### PR TITLE
Reduce code duplication between Array.forEach, some, and every

### PR DESCRIPTION
**Pull request by mmatyas**
Their implementation only differs in the stop condition and the
final return value.

JerryScript-DCO-1.0-Signed-off-by: Mátyás Mustoha mmatyas@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 1]
```diff
@@ -1890,19 +1890,49 @@ ecma_builtin_array_prototype_object_last_index_of (ecma_value_t this_arg, /**< t
...
 /**
- * The Array.prototype object's 'every' routine
- *
- * See also:
- *          ECMA-262 v5, 15.4.4.16
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
+ * If the argument is convertible to boolean true, returns ECMA_VALUE_TRUE,
+ * otherwise returns ECMA_VALUE_EMPTY.
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: To specify the return value we usually use the `@returns` element in the docs. (please see a method above)

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 2]
```diff
@@ -1890,19 +1890,49 @@ ecma_builtin_array_prototype_object_last_index_of (ecma_value_t this_arg, /**< t
...
  */
 static ecma_value_t
-ecma_builtin_array_prototype_object_every (ecma_value_t this_arg, /**< this argument */
-                                           ecma_value_t arg1, /**< callbackfn */
-                                           ecma_value_t arg2) /**< thisArg */
+ecma_builtin_convert_true_maybe (ecma_value_t var)
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: No comment for the input argument.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 3]
```diff
@@ -1890,19 +1890,49 @@ ecma_builtin_array_prototype_object_last_index_of (ecma_value_t this_arg, /**< t
...
+  }
+
+  return ECMA_VALUE_EMPTY;
+} /* ecma_builtin_convert_true_maybe */
+
+/**
+ * If the argument is convertible to boolean false, returns ECMA_VALUE_FALSE,
+ * otherwise returns ECMA_VALUE_EMPTY.
+ */
+static ecma_value_t
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: The comments are the same as above.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 4]
```diff
@@ -1890,19 +1890,49 @@ ecma_builtin_array_prototype_object_last_index_of (ecma_value_t this_arg, /**< t
...
+  }
+
+  return ECMA_VALUE_EMPTY;
+} /* ecma_builtin_convert_false_maybe */
+
+/**
+ * Applies the provided function to each element of the array as long as
+ * the return value stays undefined. If the return value stays undefined
+ * after processing the whole array, a default value can be returned.
+ */
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Just like above, the `@return` should be used.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 5]
```diff
@@ -1890,19 +1890,49 @@ ecma_builtin_array_prototype_object_last_index_of (ecma_value_t this_arg, /**< t
...
+static ecma_value_t
+ecma_builtin_array_apply (ecma_value_t this_arg, /**< this argument */
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: In this function the comments like: `/* 1. */` are referring to which part of the standard? We should mention that in the comments of the function (just like in the every/some/etc.) case.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 6]
```diff
@@ -1890,19 +1890,49 @@ ecma_builtin_array_prototype_object_last_index_of (ecma_value_t this_arg, /**< t
...
+                          ecma_value_t arg1, /**< callbackfn */
+                          ecma_value_t arg2, /**< thisArg */
+                          ecma_value_t (*check_stop_cb)(ecma_value_t), /**< retval update fn */
+                          ecma_value_t default_retval /**< default if empty */ )
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: IMHO, I would avoid using the callback method to do the work. What we could do is to only pass an enum value (something like: ARRAY_EVERY, _SOME, _FOR_EACH) and based on this value we can calculate the stop and default return values.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 7]
```diff
@@ -1890,19 +1890,49 @@ ecma_builtin_array_prototype_object_last_index_of (ecma_value_t this_arg, /**< t
...
+  {
+    return ECMA_VALUE_FALSE;
+  }
+
+  return ECMA_VALUE_EMPTY;
+} /* ecma_builtin_convert_false_maybe */
+
+/**
+ * Applies the provided function to each element of the array as long as
+ * the return value stays undefined. If the return value stays undefined
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: In the comment the `undefined` should be `empty`, as each callback returns `ECMA_VALUE_EMPTY`. The empty value is different from the undefined.

**Comment by mmatyas** [comment]
Thanks, I've updated the patch to use a routine mode enum.

**Comment by mmatyas** [comment]
Rebased to master to fix the Travis build.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 8]
```diff
@@ -1958,9 +1972,19 @@ ecma_builtin_array_prototype_object_every (ecma_value_t this_arg, /**< this argu
...
+          {
+            ret_value = ECMA_VALUE_FALSE;
+          }
+        }
+        else if (mode == ARRAY_ROUTINE_SOME)
+        {
+          if (ecma_op_to_boolean (call_value))
+          {
+            ret_value = ECMA_VALUE_TRUE;
+          }
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: This `if` structure can be merged like this: 
```c
if (mode == ARRAY_ROUTINE_EVERY || mode == ARRAY_ROUTINE_SOME)
{
    ret_value = ecma_op_to_boolean (call_value) ? ECMA_VALUE_TRUE : ECMA_VALUE_FALSE;
}
```
Therefore `ecma_op_to_boolean (call_value)` is performed only once in both cases.

Comment (mmatyas) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Ah there's a small issue actually, there are cases when the `ret_value` shouldn't change. The `if` structure can still be reduced to
```c
        if (mode == ARRAY_ROUTINE_EVERY && !ecma_op_to_boolean (call_value))
        {
            ret_value = ECMA_VALUE_FALSE;
        }
        else if (mode == ARRAY_ROUTINE_SOME && ecma_op_to_boolean (call_value))
        {
            ret_value = ECMA_VALUE_TRUE;
        }

**Pull request merged**